src/linear_algebra/general_linear_group.lean

```diff
@@ -0,0 +1,179 @@
+/-
+Copyright (c) 2021 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck
+-/
+import linear_algebra.matrix
+import linear_algebra.matrix.nonsingular_inverse
+import linear_algebra.special_linear_group
+import linear_algebra.determinant
+
+/-!
+# The General Linear group $GL(n, R)$
+
+This file defines the elements of the General Linear group `general_linear_group n R`,
+consisting of all invertible `n` by `n` `R`-matrices.
+
+
+## Main definitions
+
+ * `matrix.general_linear_group` is the type of matrices over R which are units in the matrix ring.
+ * `matrix.GL_plus.GL_pos` gives the subgroup of matrices with
+ positive determinant (over a linear ordered ring)
+
+## Implementation notes
+
+
+## References
+
+
+## Tags
+
+matrix group, group, matrix inverse
+-/
+
+namespace matrix
+universes u v
+open_locale matrix
+open linear_map
+
+section
+
+variables (n : Type u) [decidable_eq n] [fintype n] (R : Type v) [comm_ring R]
+
+/-- `GL n R` is the group of `n` by `n` `R`-matrices with unit determinant.
+Defined as a subtype of matrices-/
+abbreviation general_linear_group (R : Type*) [comm_ring R] : Type* := units (matrix n n R)
+
+notation `GL` := general_linear_group
+
+lemma GL_is_unit_det  (A : GL n R) : is_unit ((A : matrix n n R).det) :=
+begin
+  rw ←matrix.is_unit_iff_is_unit_det,
+  exact units.is_unit _
+end
```

Comment:
You can state a much stronger version of this,

```suggestion
/-- The determinant of a unit matrix is itself a unit. -/
@[simps]
def general_linear_group.det : GL n R →* units R :=
{ to_fun := λ A,
  { val := (↑A : matrix n n R).det,
    inv := (↑(A⁻¹) : matrix n n R).det,
    val_inv := by rw [←det_mul, ←mul_eq_mul, A.mul_inv, det_one],
    inv_val := by rw [←det_mul, ←mul_eq_mul, A.inv_mul, det_one]},
  map_one' := units.ext det_one,
  map_mul' := λ A B, units.ext $ det_mul _ _ }
```

Note that `simps` generates some nice lemmas for you, `↑(⇑det A) = ↑A.det`, and `↑(⇑det A)⁻¹ = ↑A⁻¹.det`

Reply:
Ok so I changed this, but it now gives a type mismatch error at `det_not_zero`. And also other type errors in the mk' which I fixed.

Reply:
Oh do I now need to call it as `det _ _ A`?

Reply:
Yes, that worked. I'm still not sure how to get out from this the fact that  `is_unit ↑A.det`. Is there an easy way to do this directly or does it need another lemma?

Reply:
The reason I mention it is that now `inv_det_pos` has a sorry as it needs this result.